lib/RAST_SDK/RAST_SDKImpl.pm

```diff
@@ -1454,143 +1454,145 @@ sub annotate_genomes
-			
-		eval {
-			my ($output,$message) = $self->annotate_process($currentparams);
-			push(@$output_genomes,$output->{ref});
-			$htmlmessage .= $message;
-		};
-		if ($@) {
-			$htmlmessage .= $input." failed!\n\n";
-		} else {
-			$htmlmessage .= $input." succeeded!\n\n";
-		}
-	}
+    my $output_genomes = [];
+    for (my $i=0; $i < @{$genomes}; $i++) {
+        my $obj_type = '';
+        my $input = $genomes->[$i];
+        if ($input =~ m/\//) {
+            my $array = [split(/\//,$input)];
+            my $info = Bio::KBase::kbaseenv::get_object_info([{ref=>
+                Bio::KBase::kbaseenv::buildref($array->[0],$array->[1],$array->[2])}],0);
+            $input = $info->[0]->[1];
+            $obj_type = $info->[0]->[2];
+        } else {
+            my $info = Bio::KBase::kbaseenv::get_object_info([{ref=>$input}]);
+            $obj_type = $info->[0]->[2];
+        }
+
+        my $currentparams = Bio::KBase::utilities::args({},[],{
+            output_genome => $input.".RAST",
+            input_genome => $genomes->[$i],
+            input_contigset => undef,
+            genetic_code => 11,
+            domain => "Bacteria",
+            scientific_name => "unknown taxon"
+        });
+        my $list = [qw(
+                    workspace
+                    scientific_name
+                    ncbi_taxon_id
+                    relation_engine_timestamp_ms
+                    genetic_code
+                    domain
+                    call_features_rRNA_SEED
+                    call_features_tRNA_trnascan
+                    call_selenoproteins
+                    call_pyrrolysoproteins
+                    call_features_repeat_region_SEED
+                    call_features_strep_suis_repeat
+                    call_features_strep_pneumo_repeat
+                    call_features_crispr
+                    call_features_CDS_glimmer3
+                    call_features_CDS_prodigal
+                    annotate_proteins_kmer_v2
+                    kmer_v1_parameters
+                    annotate_proteins_similarity
+                    resolve_overlapping_features
+                    call_features_prophage_phispy
+                    retain_old_anno_for_hypotheticals)];
+
+        for (my $j=0; $j < @{$list}; $j++) {
+            $currentparams->{$list->[$j]} = $params->{$list->[$j]} if (exists $params->{$list->[$j]});
+        }
+
+        if ($obj_type =~ /KBaseGenomeAnnotations\.Assembly/) {
+            $currentparams->{'input_contigset'} = delete $currentparams->{'input_genome'};
+            delete $currentparams->{'retain_old_anno_for_hypotheticals'};
+        }
+
+        eval {
+            my ($output,$message) = $self->annotate_process($currentparams);
+            push(@$output_genomes,$output->{ref});
+            $htmlmessage .= $message;
+        };
+        if ($@) {
+            $htmlmessage .= $input." failed!\n\n";
+        } else {
+            $htmlmessage .= $input." succeeded!\n\n";
+        }
```

Comment:
(Pre-existing code)

Presumably `$self->annotate_process` will die if something goes wrong, so this could be rewritten as

```
{
  local $@; # localise $@ to this eval block
  eval {
    my ( $output, $message ) = $self->annotate_process( $currentparams );
    push @$output_genomes, $output->{ ref };
    $htmlmessage .= $message . "$input succeeded!\n\n";
  };
  $htmlmessage .= "$input failed!\n\n" if $@;
}
```